raku-advent-2020/local-drafts/p6steve-rudolph-on-raku.md

```diff
@@ -0,0 +1,183 @@
+	has Str   $.compass is rw where <N S>.any;
+}
+class Longitude is NavAngle is export {
+	has Real  $.value is rw where 0 <= * <= 180; 
+	has Str   $.compass is rw where <E W>.any;
+}
+```
+See how the constraints are adjusted to reflect the limits of each child class.
+
+Now Rudolph can set his position:
+
+```my $lat1 = Latitude.new( value => 45, compass => <N> ); say ~$lat1; #OUTPUT 43° N```
+
+But this is quite long winded and he is impatient to get home. Raku unicode and operator overrides come
+to the rescue:
+```
+#define the pisces operator to declare and initialise a new instance 
+multi infix:<♓️> ( Any:U $left is rw, Str:D $right ) is equiv( &infix:<=> ) is export {
+    $left = NavAngle.new( $right );
+}
+```
+Now he can quickly hoof in his coordinates:
+```
+my $lat  ♓️ <55°30′30″S>; say ~$lat;   #OUTPUT 55°30.5 S
+my $long ♓️ <45°W>;       say ~$long;  #OUTPUT 45° W
+```
+## Magnetic vs. True North
+Now he knows where he is, Rudolph can set a course to steer home to the North Pole. But wait, how can
+he adjust for the difference between Magnetic north on his Compass and True North, his destination?
+
+Rudolph has another trick up his (antler) sleeve:
+```
+class CompassAdjustment { ... }
+
+our $variation = 0;			#optional variation (Compass-Adjustment)
+our $deviation = 0;			#optional deviation (Compass-Adjustment)
+
+#| Bearing embodies the identity 'M = T + Vw', so...
+#| Magnetic = True + Variation-West [+ Deviation-West]
+class Bearing is NavAngle is export {
+	has Real  $.value is rw where 0 <= * <= 360; 
+	has Str   $.compass where <M T>.any;
+
+	method M {
+		if $.compass eq <M> { return( self ) } 
+		else { return( self + ( $variation + $deviation ) ) }
+	}
+	method T {
+		if $.compass eq <T> { return( self ) } 
+		else { return( self - ( $variation + $deviation ) ) }
+	}
+
+   sub check-same( $l, $r ) {
+	if $r ~~ CompassAdjustment { 
+		return 
+	}
+        if ! $l.compass eq $r.compass {
+            die "Cannot combine Bearings of different Types!"
+        }    
+    }  
+    method add( $r is rw ) {
+        my $l = self;
+        check-same( $l, $r );
+        $l.value += $r.value;
+ 	$l.compass( $r.compass );
+        return $l
+    }    
+    method subtract( $r is rw ) {
+        my $l = self;
+        check-same( $l, $r );
+        $l.value -= $r.value; 
+	$l.compass( $r.compass );
+        return $l
+    }
+}
+
+class CompassAdjustment is Bearing is export {
+	has Real  $.value is rw where -180 <= * <= 180; 
+
+	multi method compass {						#get compass
+		given $.value {
+			when * >= 0 { return <W>, 0 }
+			when * < 0  { return <E>, 1 }
+		}
+	}
+	multi method compass( Str $compass ) {				#set compass
+		given $compass {
+			when <W>   { }		#no-op
+			when <E>   { $.value = -$.value }
+			default    { die "Compass-Adjustment must be <W E>.any" }
+		}
+	}
+}
+```
+Now, after setting the compass variation, Rudolph can enter in their magnetic compass reading and get back the 
```

Comment:
More explanations are needed here: rakudocs, when, given... Remember, we want to catch new users, if possible.

Reply:
On this part, I think that the links to rakudocs have been given earlier on - and I kind of want to retain a "bringing it all together" feel... so a bit more constrained here

Reply:
You might want to try some comments then... it's a big enough chunk of code to need a few more explanations, if possible. But anyway, it's your call.

Reply:
I have commented the code a bit in line with your suggestion...